_posts/2020-09-28-biased-locking-help.adoc

```diff
@@ -0,0 +1,84 @@
+---
+layout: post
+title: Biased locking removed from Java - does it affect you ?
+date: 2020-09-28
+tags: help
+synopsis: In Java 15 *biased locking* is removed from the Java Virtual Machine as per JEP 374. We need your help seeing if it affects Quarkus performance in your usecase.
+author: maxandersen
+---
+
+Last week we became aware that the OpenJDK team in Java 15 have disabled so called *biased locking* in the Java virtual machine. This is change from previous years and could *potentially* have a negative impact on Java applications performance.
+
+Red Hat's own performance teams are currently running performance tests to see how it affects our Java middleware, but no amount of generic testing can reveal how this change affects real world usage.
+
+That is where you come in.
+
+We would like to get info from you on wether your application performance is affected by biased locking or not.
+
+To do so please try the following in your application performance tests:
+
+Run your Qukarus application tests/performance as you normally would with the following command line flags on Java 11 (jdk11u) and if feasible Java 15 (jdk15):
+
+*enabled*: -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0
+
+*disabled*: -XX:-UseBiasedLocking
+
+We would like to know the result of this no matter if you see a regression or not on the same Java Virtual Machine.
+
+If you are able then running it with the following scenarios would be of great help too:
+
+ . single-threaded
+ . thread count ~= hardware core count
+ . thread count ~= N * hardware core count where 8 < N < 16
```

Comment:
Is this section meaningful for a Quarkus app? For WildFly I planned to omit this kind of discussion.

Reply:
Well, if a Quarkus app or performance regression test can be run with a  user-configured number of threads then yes. That may not be an option of course.

Reply:
ah - good point; I was struggling actually wording it correctly. lets just keep it simple and ask for delta effect of running with or without biased locking enabled.

Reply:
I understand this is not a very clear recommendation for many users. However, I don' think it is a good idea for us to simply to ignore how each app uses threads and how the thread count relates to the underlying hw thread capacity. Leaving that out of the question we are asking may well make it harder for us to assess the real nature of the problem.

Of course, for many apps the number of threads that will run in a deployment  will be constant. But in others it may be a configurable parameter and that's something we really need to factor in. If there is room for variation of the number of  threads that an app might use then there is immense room for both contedned and uncontended locking by those threads to influence outcome of enabling or disabling biased locking. In such cases a single data point will often be very misleading (especially if the only data point is 1 thread).

Reply:
i've pushed a version with a little bit context and make clear its only if they are able to.

trying to avoid it being "hard" to send feedback.